mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -573,44 +705,45 @@ object KMeans {
   }
 
   /**
-   * Returns the K-means cost of a given point against the given cluster centers.
+   * @return whether a center converged or not, given the epsilon parameter.
    */
-  private[mllib] def pointCost(
-      centers: TraversableOnce[VectorWithNorm],
-      point: VectorWithNorm): Double =
-    findClosest(centers, point)._2
+  override def isCenterConverged(
+      oldCenter: VectorWithNorm,
+      newCenter: VectorWithNorm,
+      epsilon: Double): Boolean = {
+    EuclideanDistanceMeasure.fastSquaredDistance(newCenter, oldCenter) <= epsilon * epsilon
```

Comment:
Do we need to override default `isCenterConverged` here? Seems to me it is equal to the default one.

Reply:
it is not, here we compare with `epsilon * epsilon`

Reply:
Add sqrt to both sides?
```scala
Math.sqrt(EuclideanDistanceMeasure.fastSquaredDistance(newCenter, oldCenter)) <= Math.sqrt(epsilon * epsilon) = epsilon
```

The left one is just the override `distance`, isn't?

Reply:
Using `sqrt` would introduce a performance regression. This is the reason why I can't use only a function to differentiate the to distance measures, because the implementation for Euclidean distance is highly optimized and this is an optimization: avoiding sqrt can be a great performance improvement since it is an expensive operation.